CompositeByteBuf optimizations and new addFlattenedComponents method

njhill opened the pull request:
Motivation:

The `CompositeByteBuf` `discardReadBytes` and `discardReadComponents` methods are currently quite inefficient, including when there are no read components to discard. We would like to call the latter more frequently in `ByteToMessageDecoder#COMPOSITE_CUMULATOR`.

In the same context it would be beneficial to perform a "shallow copy" of a composite buffer (for example when it has a refcount > 1) to avoid having to allocate and copy the contained bytes just to obtain an "independent" cumulation.

Modifications:

- Optimize `discardReadBytes()` and `discardReadComponents()` implementations (start at first comp rather than performing a binary search for the readerIndex).
- New `addFlattenedComponents(boolean,ByteBuf)` method which performs a shallow copy if the provided buffer is also composite and avoids adding any empty buffers, plus unit test.
- Other minor optimizations to avoid unnecessary checks.

Results:

`discardReadXX` methods are faster, composite buffers can be easily appended without deepening the buffer "tree" or retaining unused components.

normanmaurer commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -414,6 +414,77 @@ public CompositeByteBuf addComponents(int cIndex, Iterable<ByteBuf> buffers) {
+                final int toIdx = Math.min(widx, component.endOffset);
+                final int len = toIdx - fromIdx;
+                if (len > 0) { // skip empty components
+                    final ByteBuf buf = component.buf, slice = component.slice, newSlice;
```
can you split this in three lines (separate assignments) ?

njhill commented:
@normanmaurer I've addressed your comment and split the fix into separate PR #8946 (and amended this PR name and commit message to reflect that). I can rebase this once the other one is in.

trustin commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -414,6 +414,77 @@ public CompositeByteBuf addComponents(int cIndex, Iterable<ByteBuf> buffers) {
+     * @param buffer the {@link ByteBuf} to add. {@link ByteBuf#release()} ownership is transferred to this
+     * {@link CompositeByteBuf}.
+     */
+    public CompositeByteBuf addFlattenedComponents(boolean increaseWriterIndex, ByteBuf buffer) {
```
I'm curious if this can be named just as `addComponents()`.

njhill replied:
@trustin agree that would be much nicer but unfortunately there are already `addComponents(...)` methods with different semantics - in those cases the components added are 1-1 with the buffer(s) provided, so for example any `CompositeByteBuf` added is just "nested" as-is rather than flattened.

Would welcome other suggestions for a better name though, I don't love `addFlattenedComponents`!

normanmaurer replied:
unfortunately I dont have any better name atm... @trustin any idea ?

trustin replied:
Hmm.. not really. Just name it as `addComponents()` and document better? 😅

normanmaurer replied:
@njhill ^^

njhill replied:
@normanmaurer sorry, I'd kind of assumed @trustin was joking :). It would be a breaking chance since it would collide with the existing `addComponents(boolean, ByteBuf ...)` method (that performs a "non-flattened" addition). And even with clear doc imo it would still be confusing to have a variety of `addComponents()` methods with different args that all do the same thing and then one which happens to behave differently.

normanmaurer replied:
true you are right.... lets keep it then.

trustin commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -414,6 +414,77 @@ public CompositeByteBuf addComponents(int cIndex, Iterable<ByteBuf> buffers) {
+     * {@link CompositeByteBuf}.
+     */
+    public CompositeByteBuf addFlattenedComponents(boolean increaseWriterIndex, ByteBuf buffer) {
+        checkNotNull(buffer, "buffers");
```
`"buffer"`

trustin commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -414,6 +414,77 @@ public CompositeByteBuf addComponents(int cIndex, Iterable<ByteBuf> buffers) {
+            return this;
+        } finally {
+            if (buffer != null) {
+                // if we did not succeed, attempt to rollback any added
```
added components?

njhill commented:
Thanks @trustin, have now rebased and addressed those comments.

normanmaurer commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -414,6 +414,77 @@ public CompositeByteBuf addComponents(int cIndex, Iterable<ByteBuf> buffers) {
+                    if (slice == null) {
+                        buf.retain();
+                    } else {
+                        slice = slice.retainedSlice(fromIdx - compOffset, len);
```
please add a comment why we need to take a slice

njhill replied:
@normanmaurer thanks for highlighting this, after thinking/looking more I think a slice actually isn't required here! In particular there should be no harm retaining the `rootParent` of `PooledSlicedByteBuf`s, it's the release that we needed to be careful about in other places. I'll update.

njhill commented:
@normanmaurer now updated, PTAL

normanmaurer commented:
@netty-bot test this please

normanmaurer commented:
@netty-bot test this please

njhill commented:
@normanmaurer I think the java12 test failures are unrelated to the changes?

normanmaurer commented:
@njhill yep... I think this test code is a bit racy... will have a look.

normanmaurer commented:
@netty-bot test this please

normanmaurer commented:
@vkostyukov it will be used here: https://github.com/netty/netty/pull/8931/files

normanmaurer commented:
@njhill thanks a lot!
